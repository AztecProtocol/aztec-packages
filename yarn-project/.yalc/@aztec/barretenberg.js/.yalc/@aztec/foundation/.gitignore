dest
node_modules
.cache
.env
*.log
*.swp
.tsbuildinfo
.pnp.cjs
.pnp.loader.mjs