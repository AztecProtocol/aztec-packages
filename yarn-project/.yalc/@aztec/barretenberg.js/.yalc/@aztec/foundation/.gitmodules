[submodule "build-system"]
	path = build-system
	url = git@github.com:AztecProtocol/build-system.git
