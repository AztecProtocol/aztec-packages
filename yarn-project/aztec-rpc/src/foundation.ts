import nodeCrypto from 'crypto';
import { Keccak } from 'sha3';

export const randomBytes = (len: number) => {
  return nodeCrypto.randomBytes(len) as Buffer;
};

export function keccak256(input: Buffer | string) {
  const inputBuf = typeof input === 'string' ? Buffer.from(input) : input;
  const hash = new Keccak(256);
  return hash.update(inputBuf).digest();
}
