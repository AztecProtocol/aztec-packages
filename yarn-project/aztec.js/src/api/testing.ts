export { AnvilTestWatcher } from '../test/anvil_test_watcher.js';
export { RollupCheatCodes } from '../test/rollup_cheat_codes.js';
export { AztecCheatCodes } from '../test/aztec_cheat_codes.js';
export { EthCheatCodes } from '@aztec/ethereum/eth-cheatcodes';
export { CheatCodes } from '../test/cheat_codes.js';
