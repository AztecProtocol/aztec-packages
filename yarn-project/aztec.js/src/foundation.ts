export function sleep(ms: number) {
  return new Promise(resolve => setTimeout(resolve, ms));
}

export class Timer {
  private start: number;

  constructor() {
    this.start = new Date().getTime();
  }

  public ms() {
    return new Date().getTime() - this.start;
  }

  public s() {
    return (new Date().getTime() - this.start) / 1000;
  }
}

export function* backoffGenerator() {
  const v = [1, 1, 1, 2, 4, 8, 16, 32, 64];
  let i = 0;
  while (true) {
    yield v[Math.min(i++, v.length - 1)];
  }
}

export async function retry<Result>(fn: () => Promise<Result>, name = 'Operation', backoff = backoffGenerator()) {
  while (true) {
    try {
      return await fn();
    } catch (err: any) {
      const s = backoff.next().value;
      if (s === undefined) {
        throw err;
      }
      console.log(`${name} failed. Will retry in ${s}s...`);
      console.log(err);
      await sleep(s * 1000);
      continue;
    }
  }
}

// Call `fn` repeatedly until it returns true or timeout.
// Both `interval` and `timeout` are seconds.
// Will never timeout if the value is 0.
export async function retryUntil<T>(fn: () => Promise<T | undefined>, name = '', timeout = 0, interval = 1) {
  const timer = new Timer();
  while (true) {
    const result = await fn();
    if (result) {
      return result;
    }

    await sleep(interval * 1000);

    if (timeout && timer.s() > timeout) {
      throw new Error(name ? `Timeout awaiting ${name}` : 'Timeout');
    }
  }
}
