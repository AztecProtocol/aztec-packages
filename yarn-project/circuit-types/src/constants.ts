export const INITIAL_L2_BLOCK_NUM = 1;
export const BLOB_SIZE_IN_BYTES = 31 * 4096;
