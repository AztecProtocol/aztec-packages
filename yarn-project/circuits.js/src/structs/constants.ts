// See aztec3/constants.hpp
// Copied here for prototyping purposes
// In future: structured serialization?
export const ARGS_LENGTH = 16; // MAX_ARGS in Noir
export const RETURN_VALUES_LENGTH = 4;
export const READ_REQUESTS_LENGTH = 4;

// "PER CALL" CONSTANTS
export const MAX_NEW_COMMITMENTS_PER_CALL = 4;
export const MAX_NEW_NULLIFIERS_PER_CALL = 4;
export const MAX_PRIVATE_CALL_STACK_LENGTH_PER_CALL = 4;
export const MAX_PUBLIC_CALL_STACK_LENGTH_PER_CALL = 4;
export const MAX_NEW_L2_TO_L1_MSGS_PER_CALL = 2;
export const MAX_PUBLIC_DATA_UPDATE_REQUESTS_PER_CALL = 4;
export const MAX_PUBLIC_DATA_READS_PER_CALL = 4;

// "PER TRANSACTION" CONSTANTS
export const MAX_NEW_COMMITMENTS_PER_TX = 16; // MAX_PRIVATE_CALL_STACK_LENGTH_PER_CALL * MAX_NEW_COMMITMENTS_PER_CALL
export const MAX_NEW_NULLIFIERS_PER_TX = 16; // MAX_PRIVATE_CALL_STACK_LENGTH_PER_CALL * MAX_NEW_NULLIFIERS_PER_CALL
export const MAX_PRIVATE_CALL_STACK_LENGTH_PER_TX = 8;
export const MAX_PUBLIC_CALL_STACK_LENGTH_PER_TX = 8;
export const MAX_NEW_L2_TO_L1_MSGS_PER_TX = 2;
export const MAX_PUBLIC_DATA_UPDATE_REQUESTS_PER_TX = 4;
export const MAX_PUBLIC_DATA_READS_PER_TX = 4;
export const MAX_NEW_CONTRACTS_PER_TX = 1;
export const MAX_OPTIONALLY_REVEALED_DATA_LENGTH_PER_TX = 4;

// ROLLUP CONSTANTS
export const NUMBER_OF_L1_L2_MESSAGES_PER_ROLLUP = 16;

// TREES RELATED CONSTANTS
export const VK_TREE_HEIGHT = 3;
export const FUNCTION_TREE_HEIGHT = 4;
export const CONTRACT_TREE_HEIGHT = 16;
export const PRIVATE_DATA_TREE_HEIGHT = 32;
export const PUBLIC_DATA_TREE_HEIGHT = 254;
export const NULLIFIER_TREE_HEIGHT = 16;
export const L1_TO_L2_MSG_TREE_HEIGHT = 16;
export const PRIVATE_DATA_TREE_ROOTS_TREE_HEIGHT = 16;
export const CONTRACT_TREE_ROOTS_TREE_HEIGHT = 16;
export const L1_TO_L2_MSG_ROOTS_TREE_HEIGHT = 16;
export const ROLLUP_VK_TREE_HEIGHT = 8; // TODO: update

// SUB-TREES RELATED CONSTANTS
export const L1_TO_L2_MSG_SUBTREE_HEIGHT = Math.ceil(Math.log2(NUMBER_OF_L1_L2_MESSAGES_PER_ROLLUP));
export const L1_TO_L2_MSG_SUBTREE_SIBLING_PATH_LENGTH = L1_TO_L2_MSG_TREE_HEIGHT - L1_TO_L2_MSG_SUBTREE_HEIGHT;

// MISC CONSTANTS
export const FUNCTION_SELECTOR_NUM_BYTES = 4;
export const MAPPING_SLOT_PEDERSEN_SEPARATOR = 4;
// sha256 hash is stored in two fields to accommodate all 256-bits of the hash
export const NUM_FIELDS_PER_SHA256 = 2;
