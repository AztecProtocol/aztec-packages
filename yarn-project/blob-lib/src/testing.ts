import { BLOBS_PER_BLOCK } from '@aztec/constants';
import { makeTuple } from '@aztec/foundation/array';
import { toBufferBE } from '@aztec/foundation/bigint-buffer';
import { Fr } from '@aztec/foundation/fields';

import { Blob } from './blob.js';
import { BlobPublicInputs, BlockBlobPublicInputs } from './blob_public_inputs.js';
import { TX_START_PREFIX, TX_START_PREFIX_BYTES_LENGTH } from './encoding.js';
import { Poseidon2Sponge, SpongeBlob } from './sponge_blob.js';

/**
 * Makes arbitrary poseidon sponge for blob inputs.
 * Note: will not verify inside the circuit.
 * @param seed - The seed to use for generating the sponge.
 * @returns A sponge blob instance.
 */
export function makeSpongeBlob(seed = 1): SpongeBlob {
  return new SpongeBlob(
    new Poseidon2Sponge(
      makeTuple(3, i => new Fr(i)),
      makeTuple(4, i => new Fr(i)),
      1,
      false,
    ),
    seed,
    seed + 1,
  );
}

/**
 * Makes arbitrary blob public inputs.
 * Note: will not verify inside the circuit.
 * @param seed - The seed to use for generating the blob inputs.
 * @returns A blob public inputs instance.
 */
export function makeBlobPublicInputs(seed = 1): BlobPublicInputs {
  return new BlobPublicInputs(
    new Fr(seed),
    BigInt(seed + 1),
    makeTuple(2, i => new Fr(i)),
  );
}

/**
 * Makes arbitrary block blob public inputs.
 * Note: will not verify inside the circuit.
 * @param seed - The seed to use for generating the blob inputs.
 * @returns A block blob public inputs instance.
 */
export function makeBlockBlobPublicInputs(seed = 1): BlockBlobPublicInputs {
  return new BlockBlobPublicInputs(makeTuple(BLOBS_PER_BLOCK, () => makeBlobPublicInputs(seed)));
}

// TODO: copied form stdlib tx effect
function encodeFirstField(length: number): Fr {
  const lengthBuf = Buffer.alloc(2);
  lengthBuf.writeUInt16BE(length, 0);
  return new Fr(
    Buffer.concat([
      toBufferBE(TX_START_PREFIX, TX_START_PREFIX_BYTES_LENGTH),
      Buffer.alloc(1),
      lengthBuf,
      Buffer.alloc(1),
      Buffer.from([1]),
      Buffer.alloc(1),
      Buffer.alloc(1),
    ]),
  );
}

/**
 * Make an encoded blob with the given length
 *
 * This will deserialise correctly in the archiver
 * @param length
 * @returns
 */
export function makeEncodedBlob(length: number): Promise<Blob> {
  return Blob.fromFields([encodeFirstField(length + 1), ...Array.from({ length: length }, () => Fr.random())]);
}

/**
 * Make an unencoded blob with the given length
 *
 * This will fail deserialisation in the archiver
 * @param length
 * @returns
 */
export function makeUnencodedBlob(length: number): Promise<Blob> {
  return Blob.fromFields([...Array.from({ length: length }, () => Fr.random())]);
}

export function makeEncodedBlobFields(fields: Fr[]): Promise<Blob> {
  return Blob.fromFields([encodeFirstField(fields.length + 1), ...fields]);
}
