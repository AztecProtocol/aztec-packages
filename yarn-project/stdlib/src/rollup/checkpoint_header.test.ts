import { PROPOSED_BLOCK_HEADER_LENGTH_BYTES } from '@aztec/constants';
import { setupCustomSnapshotSerializers } from '@aztec/foundation/testing';
import { updateInlineTestData } from '@aztec/foundation/testing/files';

import { makeCheckpointHeader } from '../tests/factories.js';
import { CheckpointHeader } from './checkpoint_header.js';

describe('CheckpointHeader', () => {
  let header: CheckpointHeader;

  beforeAll(() => {
    const seed = 9870243;
    setupCustomSnapshotSerializers(expect);
    header = makeCheckpointHeader(seed);
  });

  it('serializes to buffer and deserializes it back', () => {
    const buffer = header.toBuffer();
    expect(buffer.length).toBe(PROPOSED_BLOCK_HEADER_LENGTH_BYTES);
    const res = CheckpointHeader.fromBuffer(buffer);
    expect(res).toEqual(header);
  });

  it('computes hash', () => {
    const hash = header.hash();
    expect(hash).toMatchSnapshot();
  });

  it('computes empty hash', () => {
    const header = CheckpointHeader.empty();
    const hash = header.hash();
    expect(hash).toMatchSnapshot();

    // Run with AZTEC_GENERATE_TEST_DATA=1 to update noir test data
    updateInlineTestData(
      'noir-projects/noir-protocol-circuits/crates/types/src/abis/checkpoint_header.nr',
      'test_data_empty_hash',
      hash.toString(),
    );
  });
});
