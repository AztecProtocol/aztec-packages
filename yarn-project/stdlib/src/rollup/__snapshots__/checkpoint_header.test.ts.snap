// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`CheckpointHeader computes empty hash 1`] = `Fr<0x00d72511e843bf5a2e44e8bd1da20c2626311d1d6679424f717807a1db731d62>`;

exports[`CheckpointHeader computes hash 1`] = `Fr<0x00869af74349f973964fb3bdc013c66a704cc775b54dfd425b4374f03446f881>`;
